internal/policies/apparmor/apparmor_test.go

```diff
@@ -0,0 +1,223 @@
+						&shutil.CopyTreeOptions{Symlinks: true, CopyFunction: shutil.Copy}),
+					"Setup: can't create initial apparmor dir profiles content")
+				fmt.Println(filepath.Join(apparmorDir, tc.destAlreadyExists))
+			}
+			if tc.readOnlyApparmorDir != "" {
+				testutils.MakeReadOnly(t, filepath.Join(apparmorDir, tc.readOnlyApparmorDir))
+			}
+			sat := sat{err: tc.saveAssetsError}
+
+			m := apparmor.New(apparmorDir, apparmor.WithApparmorParserCmd(apparmorParserCmd))
+			err := m.ApplyPolicy(context.Background(), "ubuntu", !tc.user, tc.entries, sat.mockSaveAssetsTo)
+			if tc.wantErr {
+				// We don't return here as we want to check that the apparmor
+				// dir is in the expected state even in error cases
+				require.Error(t, err, "ApplyPolicy should have failed but didn't")
+			} else {
+				require.NoError(t, err, "ApplyPolicy failed but shouldn't have")
+			}
+
+			// Restore permissions to be able to correctly compare trees
+			if tc.readOnlyApparmorDir != "" {
+				// nolint:gosec //false positive, this is a directory
+				err = os.Chmod(filepath.Join(apparmorDir, tc.readOnlyApparmorDir), 0700)
+				require.NoError(t, err, "can't chmod apparmor dir")
+			}
+			testutils.CompareTreesWithFiltering(t, apparmorDir, filepath.Join("testdata", "ApplyPolicy", "golden", "trees", testutils.NormalizeGoldenName(t, name)), update)
+
+			// Return early if we don't want to check apparmor_parser output for
+			// whatever reason (e.g. command did not execute, returned an error before etc.)
+			if tc.noParserOutput {
+				return
+			}
+
+			// Check that apparmor_parser was called with the expected arguments
+			goldPath := filepath.Join("testdata", "ApplyPolicy", "golden", testutils.NormalizeGoldenName(t, name))
+			got, err := os.ReadFile(parserCmdOutputFile)
+			require.NoError(t, err, "Can't read parser output file")
+			got = []byte(normalizeOutput(t, string(got), apparmorDir))
+			if update {
+				err = os.WriteFile(goldPath, got, 0600)
+				require.NoError(t, err, "Can't write golden file")
+			}
+			want, err := os.ReadFile(goldPath)
+			require.NoError(t, err, "Can't read golden file")
+			require.Equal(t, string(want), string(got), "Apparmor parser command output doesn't match")
+		})
+	}
+}
+
+func appendToFile(t *testing.T, path string, data []byte) {
+	t.Helper()
+
+	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	require.NoError(t, err, "Setup: Can't open file for appending")
+	defer f.Close()
+
+	_, err = f.Write(data)
+	require.NoError(t, err, "Setup: Can't write to file")
+}
+
+func mockApparmorParserCmd(t *testing.T, parserOutputFile string, args ...string) []string {
+	t.Helper()
+
+	cmdArgs := []string{"env", "GO_WANT_HELPER_PROCESS=1", os.Args[0], "-test.run=TestMockApparmorParser", "--", parserOutputFile}
+	cmdArgs = append(cmdArgs, args...)
+	return cmdArgs
+}
+
+func TestMockApparmorParser(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
```

Comment:
Niceeeeeeeeeeeeeee, I think it’s the first time you are doing one, right? I remember having presented you this pattern, but this is great to see!

Reply:
yep!